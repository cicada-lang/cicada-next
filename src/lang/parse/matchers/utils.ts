import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
