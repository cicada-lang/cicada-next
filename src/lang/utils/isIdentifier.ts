const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
