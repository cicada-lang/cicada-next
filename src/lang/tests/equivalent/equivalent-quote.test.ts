import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Quote", async () => {
  const output = await runCode(`

compute equivalent String {
    "abc"
  = "abc"
  = "abc"
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Quote -- empty string", async () => {
  const output = await runCode(`

compute equivalent String {
    ""
  = ""
  = ""
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Quote -- fail", async () => {
  await expectCodeToFail(`

compute equivalent String {
    "abc"
  = "abc"
  = "xyz"
}

`)
})
