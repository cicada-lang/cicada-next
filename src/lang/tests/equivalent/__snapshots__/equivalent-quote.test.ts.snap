// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Quote > "refl(implicit String, implicit \\"abc\\"): Equal(String, \\"abc\\", \\"abc\\")" 1`] = `"refl(implicit String, implicit \\"abc\\"): Equal(String, \\"abc\\", \\"abc\\")"`;

exports[`equivalent Quote -- empty string > "refl(implicit String, implicit \\"\\"): Equal(String, \\"\\", \\"\\")" 1`] = `"refl(implicit String, implicit \\"\\"): Equal(String, \\"\\", \\"\\")"`;

exports[`equivalent Quote -- empty string: "refl(implicit String, implicit ""): Equal(String, "", "")" 1`] = `"refl(implicit String, implicit ""): Equal(String, "", "")"`;

exports[`equivalent Quote: "refl(implicit String, implicit "abc"): Equal(String, "abc", "abc")" 1`] = `"refl(implicit String, implicit "abc"): Equal(String, "abc", "abc")"`;

exports[`equivalent Quote 1`] = `"refl(implicit String, implicit "abc"): Equal(String, "abc", "abc")"`;

exports[`equivalent Quote -- empty string 1`] = `"refl(implicit String, implicit ""): Equal(String, "", "")"`;
